compiler/hash-session/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+use hash_types::storage::{GlobalStorage, LocalStorage, TyStorage};
+use hash_untyped_semantics::SemanticAnalysisCtx;
+use hash_vm::InterpreterCtx;
+
+/// The [CompilerSession] holds all the information and state of the compiler
+/// instance. Each stage of the compiler contains a `State` type parameter which
+/// the compiler stores so that incremental executions of the compiler are
+/// possible.
+pub struct CompilerSession {
+    /// The collected workspace sources for the current job.
+    pub workspace: Workspace,
+    /// Any diagnostics that were collected from any stage
+    pub diagnostics: Vec<Report>,
+
+    /// The shared compiler thread pool.
+    pub pool: rayon::ThreadPool,
+
+    /// Compiler settings that are stored.
+    pub settings: CompilerSettings,
+
+    /// Sources that have passed from the `expansion` stage of the compiler.
+    /// @@Todo: Use bit-flags to represent which module has been
+    /// expanded/desugared/semantically checked/type checked.
+    pub expanded_sources: HashSet<SourceId>,
+
+    /// Sources that have passed from the `desugaring` stage of the compiler.
+    pub desugared_modules: HashSet<SourceId>,
+
+    /// Modules that have already been semantically checked. This is needed in
+    /// order to avoid re-checking modules on re-evaluations of a workspace.
+    pub semantically_checked_modules: HashSet<SourceId>,
+
+    /// Compiler type storage. Stores all the types that are created during
+    /// the typechecking stage, which is used for later stages during code
+    /// generation.
+    pub ty_storage: TyStorage,
+}
+
+impl CompilerSession {
+    pub fn new(workspace: Workspace, pool: rayon::ThreadPool, settings: CompilerSettings) -> Self {
+        let global = GlobalStorage::new();
+        let local = LocalStorage::new(&global, SourceId::default());
+
+        Self {
+            workspace,
+            diagnostics: Vec::new(),
+            pool,
+            settings,
+            ty_storage: TyStorage { global, local },
+            expanded_sources: HashSet::new(),
+            desugared_modules: HashSet::new(),
+            semantically_checked_modules: HashSet::new(),
+        }
+    }
+}
+
+impl CompilerInterface for CompilerSession {
+    fn settings(&self) -> &CompilerSettings {
+        &self.settings
+    }
+
+    fn diagnostics(&self) -> &[Report] {
+        &self.diagnostics
+    }
+
+    fn diagnostics_mut(&mut self) -> &mut Vec<Report> {
+        &mut self.diagnostics
+    }
+
+    fn workspace(&self) -> &Workspace {
+        &self.workspace
+    }
+
+    fn workspace_mut(&mut self) -> &mut Workspace {
+        &mut self.workspace
+    }
+
+    fn node_map(&self) -> &NodeMap {
+        &self.workspace.node_map
+    }
+
+    fn source_map(&self) -> &SourceMap {
+        &self.workspace.source_map
+    }
+}
+
+impl ParserCtx for CompilerSession {
+    fn data(&mut self) -> (&mut Workspace, &rayon::ThreadPool) {
+        (&mut self.workspace, &self.pool)
+    }
+}
+
```

Comment:
Maybe a function to make the compiler stages here, and import that from main/tests? Also, it might be good to typedef `Vec<Box<dyn CompilerStage<..>>>` to `CompilerStages`?